File: Day 7/Hangman.py
import random

def select_random_word(words):
    chosen_word = random.choice(words)
    return chosen_word

def get_user_guess(letters_guessed):
    while True:
        guess = input("Guess a letter: ").lower()
        if guess not in letters_guessed:
            return guess
        else:
            print(f'*****Already tried that {guess}. Pick another letter.*****\n')

def update_display(chosen_word, correct_letters):
    new_display = ""
    for letter in chosen_word:
        if letter in correct_letters:  # Check if letter is in correct letters
            new_display += letter
        else:
            new_display += "_"
    return new_display



words = ["aardvark", "camel", "hedgehog"]
stages = [''' 
  +---+
  |   |
  O   |
 /|\  |
 / \  |
      |
=======
''', '''
  +---+
  |   |
  O   |
 /|\  |
 /    |
      |
=======
''', '''
  +---+
  |   |
  O   |
 /|\  |
      |
      |
=======
''', '''
  +---+
  |   |
  O   |
 /|   |
      |
      |
=======
''', '''
  +---+
  |   |
  O   |
  |   |
      |
      |
=======
''', '''
  +---+
  |   |
  O   |
      |
      |
      |
=======
''', '''
  +---+
  |   |
      |
      |
      |
      |
=======
''']
gameOver = False
correct_letters = []
letters_guessed = []
lives = 6
chosen_word = select_random_word(words)
display = "_" * len(chosen_word)


print("Welcome to Hangman!")

while not gameOver:
    print(f'*********************You have {lives}/6 lives left*********************')
    print(letters_guessed)
    print(stages[lives])
    print(display)
    guess = get_user_guess(letters_guessed)
    if guess in letters_guessed:
        print(f'You already guessed: {guess}')
        continue
    letters_guessed.append(guess)


    if guess in chosen_word:
        correct_letters.append(guess)
        display = update_display(chosen_word, correct_letters)
    else:
        lives -= 1
        if lives == 0:
            print(stages[0])
            print("You lose! The word was:", chosen_word)
            gameOver = True

    
    if "_" not in display:
        print(f'The word was {chosen_word}, good job.')
        print("**************************You win**************************")
        gameOver = True

File: Day 7/Hangman.java
import java.util.HashSet;
import java.util.Random;
import java.util.Scanner;

public class Hangman {
    private static final String[] WORDS = {"aardvark", "camel", "hedgehog"};
    private static final int MAX_LIVES = 6;

    public static void main(String[] args) {
        Scanner userInput = new Scanner(System.in);
        String chosenWord = selectRandomWord();
        StringBuilder display = new StringBuilder("_".repeat(chosenWord.length()));
        HashSet<Character> correctLetters = new HashSet<>();
        HashSet<Character> lettersGuessed = new HashSet<>();
        int lives = MAX_LIVES;
        boolean gameOver = false;

        while (!gameOver) {
            System.out.println("Current word: " + display);
            System.out.println("Lives left: " + lives);
            char guess = getUserGuess(userInput);

            if (lettersGuessed.contains(guess)) {
                System.out.println("You already guessed that letter.");
                continue; // Skip the rest of the loop
            }

            lettersGuessed.add(guess);
            if (chosenWord.indexOf(guess) >= 0) {
                correctLetters.add(guess);
                updateDisplay(chosenWord, display, correctLetters);
            } else {
                lives--;
            }

            if (display.toString().equals(chosenWord)) {
                System.out.println("You win! The word was: " + chosenWord);
                gameOver = true;
            } else if (lives <= 0) {
                System.out.println("You lose! The word was: " + chosenWord);
                gameOver = true;
            }
        }
        userInput.close();
    }

    private static String selectRandomWord() {
        Random random = new Random();
        return WORDS[random.nextInt(WORDS.length)];
    }

    private static char getUserGuess(Scanner userInput) {
        char guess;
        while (true) {
            System.out.print("Guess a letter: ");
            String input = userInput.nextLine();
            if (input.length() == 1 && Character.isLetter(input.charAt(0))) {
                guess = Character.toLowerCase(input.charAt(0)); // Normalize to lowercase
                break;
            } else {
                System.out.println("Please enter a valid single letter.");
            }
        }
        return guess;
    }

    private static void updateDisplay(String chosenWord, StringBuilder display, HashSet<Character> correctLetters) {
        StringBuilder newDisplay = new StringBuilder();
        for (char letter : chosenWord.toCharArray()) {
            if (correctLetters.contains(letter)) {
                newDisplay.append(letter);
            } else {
                newDisplay.append('_');
            }
        }
        display.setLength(0); // Clear the current display
        display.append(newDisplay); // Update display with the new state
    }
}
